DOCSP-30510: Find One Usage Example (#68)
mongodb/docs-rust: source/usage-examples/findOne.txt

.. _rust-find-one-usage:

===============
Find a Document
===============

You can retrieve a single document in a collection by calling the `find_one() 
<{+api+}/struct.Collection.html#method.find_one>`__ method on a ``Collection`` instance. 

Pass a query filter to the ``find_one()`` method to return one document in the collection
that matches the filter. If multiple documents match the query filter, this method returns

> the first matching document according to their :term:`natural order` in the database, or

Reviewer: Suggestion:

I think commas before "or" should only be used when it separates an independent clause. Since the phrase starting with "according to the sort order..." isn't an independent clause, it should be omitted.

```suggestion
the first matching document according to their :term:`natural order` in the database or
```

Author: I think I would still need a comma here, since the phrase following the comma means the same as what precedes (and "or" is used to introduce a synonym/explanation)

Reviewer: I think "or" is a conjunction that signals a different item, not a synonym or explanation. When presenting two different items, it should not be preceded by a comma. Let me know if I'm missing something.

according to the sort order specified in a ``FindOneOptions`` instance. To learn more
about natural order, see :manual:`natural order </reference/glossary/#std-term-natural-order>`

> in the Server manual glossary.

Reviewer: Issue:
This sentence repeats the link in the previous sentence.

Suggestion:

I think one of the following actions could resolve this issue:
- Omit the "term" role in the previous section.
- Omit the last sentence since the "term" role in the previous sentence contains the same link.

The ``find_one()`` method returns a `Result<Option<T>> <{+api+}/error/type.Result.html>`__
type. 

.. tip::

    To learn more about retrieving documents, see the :ref:`rust-retrieve-guide` guide.

Example
-------

This example retrieves a document from the ``restaurants`` collection in the
``sample_restaurants`` database that matches a query filter. The example populates a

> ``Restaurant`` struct with the data from the retrieved document.

Reviewer: Suggestion:

I think the description of what document is retrieved (matches the filter) should be placed next to the mention of the document rather than separated by the description of where the document is retrieved from. E.g.

"This example retrieves a document that matches a specified query filter from..."

Similar suggestion for the find multiple documents usage example (as well as [this comment](https://github.com/mongodb/docs-rust/pull/68#discussion_r1379390868)).

.. _rust-find-one-usage:

===============
Find a Document
===============

You can retrieve a single document in a collection by calling the ``find_one()``
method on a ``Collection`` instance. 

Pass a query filter to the ``find_one()`` method to return one document in the collection
that matches the filter. If multiple documents match the query filter, the ``find_one()``
method returns the first matching document according to the documents' natural order, or as
the documents appear in the database.

The ``find_one()`` method returns a ``Result<Option<T>>`` type. This type has a value of

> ``Some`` if a document matches the query filter and ``None`` if no documents match.

Reviewer: Suggestion:

I think this info about "Some" and "None" values could be omitted since this information repeats the [Retrieve Data](https://www.mongodb.com/docs/drivers/rust/current/fundamentals/crud/read-operations/retrieve/#find-one-document) content and isn't used in the Usage Example.

Example
-------

> This example finds a document from the ``sample_restaurants.restaurants`` collection

Reviewer: Issue:

Since collection names can have periods in them, this refers to a collection with the name "sample_restaurants.restaurants" rather than what I think is intended (a database name of "sample_restauarants" and collection name of "restaurants").

Suggestion:

I think calling it a "namespace" works, but maybe identifying the database name and collection name separately would be helpful for the target personas. I think the example retrieves a document that matches a query filter. I think "finds" might be used more for cases in which there's only a check for its existence.

```suggestion
This example retrieves a document from the ``restaurants`` collection in the ``sample_restaurants`` database
```

Author: Fixed -- I think all of the other Rust usage ex pages so far have used "sample_restaurants.restaurants," so would be good to standardize that. I'll fix it in one of my open PRs.

Reviewer: Thanks. Another suggestion:

I think the description of what document is retrieved (matches the filter) should be placed next to the mention of the document rather than separated by the description of where the document is retrieved from. E.g.

"This example retrieves a document that matches a specified query filter from..."

> that matches a query filter. The example uses a ``Restaurant`` struct that has ``name``
> and ``cuisine`` fields to model the documents in the collection.

Reviewer: Suggestion:

Since modeling of documents (comparing the document to the struct) is not shown, that information could be omitted. Instead, I think it could be more helpful to identify its purpose in the example.

```suggestion
that matches a query filter. The example populates a ``Restaurant`` struct with the data from the retrieved document.
```

The following code passes a query filter as a parameter to the ``find_one()`` method.

> The filter matches documents where the value of the ``name`` field is ``"Tompkins Square Bagels"``.

Reviewer: Suggestion:

I think that this could be made more concise by assuming that the reader read the previous section that contained similar information ("Pass a query filter...").

```suggestion
The following code example uses a query filter that matches documents in which the value of the ``name`` field is ``"Tompkins Square Bagels"``.

> 

Reviewer: Question:

Would it be helpful to explain that they should select the tab that best applies to their usage, or is there some other information in the docs about this that I wasn't able to locate?

I noticed that in [Retrieve Data](https://www.mongodb.com/docs/drivers/rust/current/fundamentals/crud/read-operations/retrieve/), there isn't any information explaining it's the asynchronous version which could be confusing.

Author: I think including a sentence about selecting the right tab, similar to C# usage examples, is a good idea

                Restaurant {
                    name: "Tompkins Square Bagels",
                    cuisine: "American",
                },
            )

   .. tab:: Synchronous
      :tabid: find-sync

      .. io-code-block::
         :copyable: true

        .. input:: /includes/usage-examples/code-snippets/find-one-sync.rs
            :language: rust
            :dedent:

        .. output::
            :language: console
            :visible: false

            Some(
                Restaurant {
                    name: "Tompkins Square Bagels",
                    cuisine: "American",
                },
            )

Additional Information
----------------------

To learn more about retrieving documents, see the :ref:`rust-retrieve-guide` guide.

To learn more about natural order, see :manual:`natural order </reference/glossary/#std-term-natural-order>`
in the Server manual glossary.

To learn more about using query filters, see the :ref:`rust-query-guide` guide.

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about the methods and types mentioned on this
page, see the following API documentation:

- `find_one() <{+api+}/struct.Collection.html#method.find_one>`__

> - `Result <{+api+}/error/type.Result.html>`__

Reviewer: Suggestion:

Not sure if you're set on including these sections on every Usage Example page, but I think these repeat info in the Retrieve Data page. I think the page could be made more concise and less repetitive if these sections were replaced by the following:

- Inline API links for the first mention of the types and methods
- Include a link to the Retrieve Data page in the introduction since this information should be covered there
